node/src/components/contract_runtime/operations.rs

```diff
@@ -155,61 +131,121 @@ pub fn execute_finalized_block(
 
     Ok(BlockAndExecutionEffects {
         block,
-        execution_results,
+        execution_results: execution_results.into_iter().collect(),
         maybe_step_effect_and_upcoming_era_validators,
     })
 }
 
-/// Commits the execution effects.
-fn commit_execution_effects(
+type ExecResultMap = BTreeMap<NodeDeployHash, (DeployHeader, ExecutionResult)>;
+
+fn execute_and_commit(
     engine_state: &EngineState<LmdbGlobalState>,
     metrics: Option<Arc<ContractRuntimeMetrics>>,
+    protocol_version: ProtocolVersion,
+    finalized_block: &FinalizedBlock,
+    deploys: Vec<Deploy>,
     state_root_hash: Digest,
-    deploy_hash: DeployHash,
-    execution_results: ExecutionResults,
-) -> Result<(Digest, ExecutionResult), BlockExecutionError> {
-    let ee_execution_result = execution_results
-        .into_iter()
-        .exactly_one()
-        .map_err(|_| BlockExecutionError::MoreThanOneExecutionResult)?;
-    let json_execution_result = ExecutionResult::from(&ee_execution_result);
-
-    let execution_effect: AdditiveMap<Key, Transform> = match ee_execution_result {
-        EngineExecutionResult::Success {
-            execution_journal,
-            cost,
-            ..
-        } => {
-            // We do want to see the deploy hash and cost in the logs.
-            // We don't need to see the effects in the logs.
-            debug!(?deploy_hash, %cost, "execution succeeded");
-            execution_journal
+) -> Result<(ExecResultMap, Digest), BlockExecutionError> {
+    let block_time = finalized_block.timestamp().millis();
+
+    // Create a new EngineState that reads from LMDB but only caches changes in memory.
+    let scratch_state = engine_state.get_scratch_engine_state();
+
+    let deploy_items = deploys
+        .iter()
+        .cloned()
+        .map(DeployItem::from)
+        .collect::<Vec<_>>();
+
+    let mut execution_results = BTreeMap::new();
```

Comment:
Why not make it a `mut Vec` and push in line 191 rather than insert into a tree? We execute in order so we should collect the results in order as well.

Reply:
We don't have to clone the data - we can simply iterate over the input `deploys`. I had created a PR to your repo with that approach - it avoid cloning the data which you do twice in here.